### PR TITLE
Dirty Flag pattern #560

### PR DESCRIPTION
**Pull request by waisuan**
- Implementation of the Dirty Flag pattern

### CODE REVIEW

**Comment by npathai** [comment]
@iluwatar I will pick up this review.

**Comment by waisuan** [comment]
@iluwatar Done.

**Comment by waisuan** [comment]
Seems to be a JUnit test failure:-

dummyCustomerApiTest()  Time elapsed: 0.052 sec  <<< FAILURE!
org.opentest4j.AssertionFailedError: Counter limit must be reached ==> expected: <true> but was: <false>
	at com.iluwatar.throttling.B2BServiceTest.dummyCustomerApiTest(B2BServiceTest.java:46)

Doesn't seem to be related to my changes here.

@iluwatar

**Comment by iluwatar** [comment]
It is a known issue https://github.com/iluwatar/java-design-patterns/issues/643, still unresolved

**Comment by iluwatar** [comment]
@npathai please review the PR

**Comment by npathai** [comment]
@iluwatar Yes started. Will complete today

File: dirty-flag/src/main/java/com/iluwatar/dirtyflag/World.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  private World() {
+  }
+
+  /**
+   * Init.
+   * 
+   * @return World instance
+   */
+  public static World getInstance() {
+    if (world == null) {
```
Comment (npathai) [inline dirty-flag/src/main/java/com/iluwatar/dirtyflag/World.java]: Suggestion: Just document a warning that this is not a thread safe way to implement Singleton. Refer to Singleton for more information.

File: dirty-flag/src/main/java/com/iluwatar/dirtyflag/DataFetcher.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  private DataFetcher() {
+  }
+
+  /**
+   * Init.
+   * 
+   * @return DataFetcher instance
+   */
+  public static DataFetcher getInstance() {
```
Comment (npathai) [inline dirty-flag/src/main/java/com/iluwatar/dirtyflag/DataFetcher.java]: Question: I don't understand the intent to keep `World` and `DataFetcher` singleton here. Any particular reason you chose this way?

File: dirty-flag/src/test/java/org/dirty/flag/DirtyFlagTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ *
+ * Application test
+ *
+ */
+public class DirtyFlagTest {
+
+  @BeforeEach
+  public void reset() throws SecurityException, NoSuchFieldException, IllegalArgumentException, IllegalAccessException {
+    Field instance = DataFetcher.class.getDeclaredField("df");
```
Comment (npathai) [inline dirty-flag/src/test/java/org/dirty/flag/DirtyFlagTest.java]: This singleton reset thing won't be needed if we avoid `getInstance`. Hence my question about Singletonness.

File: dirty-flag/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ - Java
+ - Difficulty-Easy
+ - Performance
+---
+
+## Intent
+To avoid expensive re-acquisition of resources. The resources retain their identity, are kept in some
+fast-access storage, and are re-used to avoid having to acquire them again.
+
+![alt text](./etc/dirty-flag.png "Dirty Flag")
```
Comment (npathai) [inline dirty-flag/README.md]: The class diagram should be made with ObjectAid eclipse plugin. More about it in wiki

**Comment by npathai** [comment]
@waisuan You have my review comments. Overall its looks great and easy to follow. There are a few points that I have highlighted though. Awaiting your response.

**Comment by npathai** [comment]
@waisuan Any updates?

**Comment by waisuan** [comment]
@npathai Apologies for the delay here. Been swamped. I'll take a look at your comments tonight and begin working on them accordingly. Cheers!

File: dirty-flag/README.md (hunk 4 above)
Comment (waisuan) [reply dirty-flag/README.md]: Done.

**Comment by waisuan** [comment]
@npathai -- Fixes are up. Could you take another peek? Thanks!

**Comment by npathai** [comment]
@waisuan :+1: Surely. I will do that tomorrow.

File: dirty-flag/src/test/java/org/dirty/flag/DirtyFlagTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ * Application test
+ *
+ */
+public class DirtyFlagTest {
+
+  @Test
+  public void testIsDirty() {
+    DataFetcher df = new DataFetcher();
+    List<String> countries = df.fetch();
+    assertTrue(!countries.isEmpty());
```
Comment (npathai) [inline dirty-flag/src/test/java/org/dirty/flag/DirtyFlagTest.java]: Use `assertFalse()` in place of `assertTrue(!condition)`

**Pull request merged**